File: services/backend-for-frontend-python/o11yday_lib.py
import requests

SERVICES = {
    'meminator': 'http://meminator:10114/applyPhraseToPicture', 
    'phrase-picker': 'http://phrase-picker:10114/phrase',
    'image-picker': 'http://image-picker:10114/imageUrl',
}

def fetch_from_service(service, method='GET', body=None):
    """
    Fetches data from a remote service over HTTP.

    Args:
        service (str): One of the services in SERVICES
        method: 'GET' or 'POST'
    Returns:
        dict: The JSON response from the remote service.
    """
    try:
        url = SERVICES[service] # what happens if it is not there?
        if method == 'GET':
            response = requests.get(url)
        elif method == 'POST':
            response = requests.post(url, json=body)
        else:
            print(f"Method {method} not supported")
        response.raise_for_status()  # Raise an exception for 4xx or 5xx status codes
        return response # do not parse the response here
    except requests.RequestException as e:
        print(f"Error fetching data from {url}: {e}")
        return None

File: services/backend-for-frontend-deno/src/o11yday-lib.ts
/**
 * This represents an internal library used for communication between services.
 * That is a common pattern, and a great place to customize some telemetry.
 */

import { context, trace, Attributes, SpanStatusCode, Span, SpanOptions } from '@opentelemetry/api';

const SERVICES = {
    meminator: 'http://meminator:10116/applyPhraseToPicture', // this one is a POST
    'phrase-picker': 'http://phrase-picker:10117/phrase',
    'image-picker': 'http://image-picker:10118/imageUrl',
}

type FetchOptions = {
    method?: "GET" | "POST" | "PUT" | "DELETE",
    body?: any // will be stringified
}

/**
 * Make an HTTP request to one of our services.
 * 
 * INSTRUMENTATION: internal libraries like this one are a great place to add telemetry.
 * 
 * @param service one of our known services
 * @param options choose method and/or send a body
 * @returns 
 */
export async function fetchFromService(service: keyof typeof SERVICES, options?: FetchOptions) {
    const { method, body: bodyObject } = options || { method: "GET" };
    let body: string | null = null;
    if (bodyObject) {
        body = JSON.stringify(bodyObject);
    }
    const headers: Record<string, string> = {
        "Content-Type": "application/json"
    }

    const url = SERVICES[service];

    const response = await fetch(url, { headers, method, body });

    return response;
}
